### PR TITLE
Support exclusive allocation on non-exclusive host

### PR DESCRIPTION
**Pull request by hakonhall**
This PR is a no-op.  But when enabling the new make-exclusive flag:
- An exclusive allocation will be allowed on a non-exclusive host
- The host is then made exclusive

### CODE REVIEW

**Comment by hakonhall** [comment]
FYI @jonmv

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java [hunk 1]
```diff
@@ -186,18 +192,15 @@ private boolean checkForClashingParentHost() {
...
     private boolean offeredNodeHasParentHostnameAlreadyAccepted(NodeCandidate candidate) {
-        for (NodeCandidate acceptedNode : nodes.values()) {
-            if (acceptedNode.parentHostname().isPresent() && candidate.parentHostname().isPresent() &&
-                    acceptedNode.parentHostname().get().equals(candidate.parentHostname().get())) {
-                return true;
-            }
-        }
-        return false;
+        if (candidate.parentHostname().isEmpty()) return false;
+        return nodes.values().stream().anyMatch(acceptedNode -> acceptedNode.parentHostname().equals(candidate.parentHostname()));
```
Comment (jonmv) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java:196]: 👍

**Review by jonmv — APPROVED** [review]
LGTM!

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/applications/Applications.java [hunk 2]
```diff
@@ -63,19 +63,24 @@ public void remove(ApplicationTransaction transaction) {
         db.deleteApplication(transaction);
     }
 
+    public record Lock(Mutex mutex, ApplicationId application) implements Mutex {
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/applications/Applications.java:66]: FYI: There's already a type for this: `com.yahoo.config.provision.ProvisionLock`

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/HostCapacityMaintainerTest.java [hunk 3]
```diff
@@ -316,6 +320,44 @@ public void respects_exclusive_allocation() {
         tester.assertNodesUnchanged();
     }
 
+    @Test
+    public void works_as_before_without_make_exclusive() {
```
Comment (freva) [inline node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/HostCapacityMaintainerTest.java:324]: What about a test with `make-exclusive`?

Comment (jonmv) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/HostCapacityMaintainerTest.java:324]: The above test has that.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/applications/Applications.java (hunk 2 above)
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/applications/Applications.java:66]: Thanks, I failed to find such a class even though I vaguely remembered there was one.  Fixed in https://github.com/vespa-engine/vespa/pull/29070

**Pull request merged**